auto_round/export/export_to_awq.py

```diff
@@ -0,0 +1,251 @@
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import torch
+import torch.nn as nn
+
+from auto_round.export.register import register_format
+from auto_round.utils import check_to_quantized, convert_dtype_torch2str_hf, get_block_names, get_module, logger
+
+# MIT License
+
+# Copyright (c) 2023 MIT HAN Lab
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+
+@register_format("auto_awq")
+def save_quantized_as_autoawq(output_dir, model_path, inplace=True, **kwargs):
+    """Export the model to autogptq format to easily leverage cuda kernel."""
+    model = kwargs["model"]
+    weight_config = kwargs["weight_config"]
+    sym = kwargs["sym"]
+    bits = kwargs["bits"]
+    group_size = kwargs["group_size"]
+    iters = kwargs["iters"]
+    lr = kwargs["lr"]
+    minmax_lr = kwargs["minmax_lr"]
+    enable_minmax_tuning = kwargs["enable_minmax_tuning"]
+    enable_quanted_input = kwargs["enable_quanted_input"]
+    scale_dtype = kwargs["scale_dtype"]
+    tokenizer = kwargs["tokenizer"]
+    supported_types = kwargs["supported_types"]
+
+    logger.info("Saving quantized model to autoawq format")
+    if tokenizer is not None:
+        tokenizer.save_pretrained(output_dir)
+    ##check module quantized in block, this may have bug for mixed precision quantization
+    block_name = get_block_names(model)[0]
+    first_block = get_module(model, block_name)
+    all_to_quantized = True
+    modules_in_block_to_quantize = []
+    for n, m in first_block.named_modules():
+        is_supported_type = False
+        for supported_type in supported_types:
+            if isinstance(m, supported_type):
+                is_supported_type = True
+                break
+        if not is_supported_type:
+            continue
+        if not check_to_quantized(m):
+            all_to_quantized = False
+        else:
+            modules_in_block_to_quantize.append(n)
+    modules_in_block_to_quantize = [modules_in_block_to_quantize]
+    if all_to_quantized:
+        modules_in_block_to_quantize = None
```

Comment:
this may have bug for mixstral moe model, as in main.py, some gate layer are excluded